clients/src/main/java/org/apache/kafka/clients/producer/ProducerInterceptor.java

```diff
@@ -0,0 +1,80 @@
+ * by KafkaProducer if not specified in the producer config. The interceptor implementation needs to be aware that it will be
+ * sharing producer config namespace with other interceptors and serializers, and ensure that there are no conflicts.
+ * <p>
+ * ProducerInterceptor callbacks may be called from multiple threads. Interceptor implementation must ensure thread-safety, if needed.
+ */
+public interface ProducerInterceptor<K, V> extends Configurable {
+    /**
+     * This is called when client sends the record to KafkaProducer, before key and value gets serialized.
+     * <p>
+     * This method is allowed to mutate the record, in which case, the new record will be returned. The implication of modifying
+     * key/value is that partition assignment (if not specified in ProducerRecord) will be done based on modified key/value,
+     * not key/value from the client. As a result, key and value transformation done in onSend() needs to be consistent:
+     * same key and value should mutate to the same (modified) key and value. Otherwise, log compaction would not work
+     * as expected.
+     * <p>
+     * Similarly, it is up to interceptor implementation to ensure that correct topic/partition is returned in ProducerRecord.
+     * Most often, it should be the same topic/partition from 'record'.
+     * <p>
+     * Any exception thrown by this method will be caught by the caller and logged, but not propagated further.
+     * <p>
+     * Since the producer may run multiple interceptors, a particular interceptor's onSend() callback will be called in order specified in
+     * configuration {@link org.apache.kafka.clients.producer.ProducerConfig#INTERCEPTOR_CLASSES_CONFIG}. The first interceptor
+     * in the list gets the record passed from the client. The following interceptor will be passed the record returned by the
+     * previous interceptor, and so on. Since interceptors are allowed to mutate records, interceptors may potentially get
+     * the record already mutated by other interceptors. However, building a pipeline of mutable interceptors that depend on the output
+     * of the previous interceptor is discouraged, because of potential side-effects caused by interceptors potentially failing to mutate the record
+     * and throwing an exception. If one of the interceptors in the list throws an exception from onSend(), the exception is caught, logged, and
+     * the next interceptor is called with the record returned by the last successful interceptor in the list, or otherwise the client.
+     *
+     * @param record the record from client or the record returned by the previous interceptor in the chain of interceptors.
+     * @return producer record to send to topic/partition
+     */
+    public ProducerRecord<K, V> onSend(ProducerRecord<K, V> record);
+
+    /**
+     * This method is called when the record sent to the server has been acknowledged, or when sending the record fails before
```

Comment:
I find the usage of the word `when` a bit imprecise in this context. Would it not be better to use `before` or `after` (potentially qualified with `just`, `immediately` and such things)? I think there are quite a few cases of this, so again, it's probably worth getting a few opinions before changing (unless you are sure that my suggestion is an improvement, of course).